src/ds++/SP.hh

```diff
@@ -13,321 +13,14 @@
- */
-template <typename T>
-template <class U>
-SP<T>::SP(U *p) : p_(p), r_(new SPref) {
-  Ensure(get() == p);
-  Ensure(unique());
-}
-
-//---------------------------------------------------------------------------//
-/*!
- * \brief Copy constructor for SP<T>.
- *
- * \param sp_in smart pointer of type SP<T>
- */
-template <typename T>
-SP<T>::SP(const SP<T> &sp_in) : p_(sp_in.p_), r_(sp_in.r_) {
-  // advance the reference to T if not empty
-  if (r_)
-    r_->refs++;
-}
+template <class T> using SP = std::shared_ptr<T>;
 
-//---------------------------------------------------------------------------//
-/*!
- * \brief Copy constructor for SP<U>.
- *
- * This copy constructor requires that U * is implicitly convertible to T *.
- *
- * \param x smart pointer of type SP<U>
- */
-template <typename T>
-template <class U>
-SP<T>::SP(const SP<U> &x) : p_(x.p_), r_(x.r_) {
-  // advance the reference to T if not empty
-  if (r_)
-    r_->refs++;
-}
-
-//---------------------------------------------------------------------------//
-/*!
- * \brief Assignment operator for type SP<T>.
- *
- * \param x smart pointer of type SP<T>
- */
-template <typename T> SP<T> &SP<T>::operator=(const SP<T> &x) {
-  SP<T>(x).swap(*this);
-  return *this;
-}
-
-//---------------------------------------------------------------------------//
-/*!
- * \brief Assignment operator for type SP<U>.
- *
- * This assignment requires that U * is implicitly convertible to T *.  The
- * pointer in x can point to NULL.
- *
- * \param x smart pointer of type SP<U>
- */
-template <typename T>
-template <class U>
-SP<T> &SP<T>::operator=(const SP<U> &x) {
-  SP<T>(x).swap(*this);
-  return *this;
-}
-
-//---------------------------------------------------------------------------//
-template <class T, class U> SP<T> dynamic_pointer_cast(const SP<U> &sp) {
-  SP<T> Result;
-  T *pt = dynamic_cast<T *>(sp.p_);
-  if (pt) {
-    Result.p_ = pt;
-    Result.r_ = sp.r_;
-    sp.r_->refs++;
-  }
-  return Result;
-}
-
-//---------------------------------------------------------------------------//
-template <class T> template <typename U> void SP<T>::reset(U *p) {
-  free();
-  p_ = p;
-  r_ = new SPref;
-}
-
-//---------------------------------------------------------------------------//
-template <class T> void SP<T>::swap(SP<T> &r) {
-  T *tp = p_;
-  p_ = r.p_;
-  r.p_ = tp;
-  SPref *rp = r_;
-  r_ = r.r_;
-  r.r_ = rp;
-}
-
-//---------------------------------------------------------------------------//
-// PRIVATE IMPLEMENTATION
-//---------------------------------------------------------------------------//
-/*!
- * \brief Decrement the count and free the pointer if count is zero.
- *
- * Note that it is perfectly acceptable to call delete on a NULL pointer.
- */
-template <typename T> void SP<T>::free() {
-  // if the count goes to zero then we free the data
-  if (r_ && --r_->refs == 0) {
-    delete p_;
-    delete r_;
-  }
-}
+using std::dynamic_pointer_cast;
```

Comment:
This type of namespace injection makes me nervous.  We should discuss this at the meeting tomorrow.